--- ../src-base/minecraft/net/minecraft/item/ItemStack.java
+++ ../src-work/minecraft/net/minecraft/item/ItemStack.java
@@ -1,5 +1,6 @@
 package net.minecraft.item;
 
+import carpet.CarpetSettings;
 import com.google.common.collect.HashMultimap;
 import com.google.common.collect.Multimap;
 import java.text.DecimalFormat;
@@ -20,6 +21,7 @@
 import net.minecraft.entity.player.EntityPlayerMP;
 import net.minecraft.init.Blocks;
 import net.minecraft.init.Enchantments;
+import net.minecraft.init.Items;
 import net.minecraft.inventory.EntityEquipmentSlot;
 import net.minecraft.nbt.NBTBase;
 import net.minecraft.nbt.NBTTagCompound;
@@ -41,6 +43,8 @@
 import net.minecraft.util.text.translation.I18n;
 import net.minecraft.world.World;
 
+import carpet.helpers.StatHelper;
+
 public final class ItemStack
 {
     public static final ItemStack field_190927_a = new ItemStack((Item)null);
@@ -98,7 +102,10 @@
 
     private void func_190923_F()
     {
-        this.field_190928_g = this.func_190926_b();
+        if (CarpetSettings.optimizedItemStackEmptyCheck)
+            this.field_190928_g = (this.field_151002_e == null || this.field_151002_e == Item.func_150898_a(Blocks.field_150350_a) || this.field_77994_a <= 0 || this.field_77991_e < -32768 || this.field_77991_e > 65535);
+        else
+            this.field_190928_g = this.func_190926_b();
     }
 
     public ItemStack(NBTTagCompound p_i47263_1_)
@@ -122,6 +129,9 @@
 
     public boolean func_190926_b()
     {
+        if (CarpetSettings.optimizedItemStackEmptyCheck)
+            return this.field_190928_g;
+
         if (this == field_190927_a)
         {
             return true;
@@ -169,7 +179,8 @@
 
         if (enumactionresult == EnumActionResult.SUCCESS)
         {
-            p_179546_1_.func_71029_a(StatList.func_188057_b(this.field_151002_e));
+            // CM add meta
+            p_179546_1_.func_71029_a(StatHelper.getObjectUseStats(this.field_151002_e, this.field_77991_e));
         }
 
         return enumactionresult;
@@ -334,7 +345,8 @@
 
         if (flag)
         {
-            p_77961_2_.func_71029_a(StatList.func_188057_b(this.field_151002_e));
+            // CM add meta
+            p_77961_2_.func_71029_a(StatHelper.getObjectUseStats(this.field_151002_e, this.field_77991_e));
         }
     }
 
@@ -344,7 +356,8 @@
 
         if (flag)
         {
-            p_179548_4_.func_71029_a(StatList.func_188057_b(this.field_151002_e));
+            // CM add meta
+            p_179548_4_.func_71029_a(StatHelper.getObjectUseStats(this.field_151002_e, this.field_77991_e));
         }
     }
 
@@ -496,7 +509,8 @@
 
     public void func_77980_a(World p_77980_1_, EntityPlayer p_77980_2_, int p_77980_3_)
     {
-        p_77980_2_.func_71064_a(StatList.func_188060_a(this.field_151002_e), p_77980_3_);
+        // CM add meta
+        p_77980_2_.func_71064_a(StatHelper.getCraftStats(this.field_151002_e, this.field_77991_e), p_77980_3_);
         this.func_77973_b().func_77622_d(this, p_77980_1_, p_77980_2_);
     }
 
@@ -881,4 +895,10 @@
     {
         this.func_190917_f(-p_190918_1_);
     }
+
+    // Check for ground stacking CARPET-XCOM
+	public boolean isGroundStackable() {
+		return this.func_77973_b().itemGroundStacking(func_77942_o());
+	}
+	public void forceStackSize(int size){this.field_77994_a = size;}
 }
